### PR TITLE
Add support for generating code coverage with Jacoco

### PR DESCRIPTION
**Pull request by andyscott**
This adds preliminary support for generating code coverage. I tried to be as lazy as possible and defer all evaluation/instrumentation until it's actually needed.

There's one issue right now-- code coverage for `java_xyz` rules included transitively doesn't show up in coverage output provided from a `scala_test`. This is documented in the tests.

And as a side note: I have almost the exact same code working for https://github.com/bazelbuild/rules_scala. I did development for both by implementing functionality here, first. Once I was happy with it I tried it in https://github.com/bazelbuild/rules_scala. The PR for that repo should land tomorrow.

### CODE REVIEW

File: rules/scala.bzl [hunk 1]
```diff
@@ -182,6 +206,7 @@ def _scala_library_implementation(ctx):
         ("javainfo", _phase_javainfo),
         ("compile", _phase_noop),
         ("singlejar", _phase_singlejar),
+        ("coverage", _phase_coverage_jacoco),
```
Comment (borkaehw) [inline rules/scala.bzl]: As far as I can tell, `bazel coverage` only works against test targets, why would we need for `scala_library`? I apologize if I am wrong.

File: rules/scala.bzl [hunk 2]
```diff
@@ -194,6 +219,7 @@ def _scala_binary_implementation(ctx):
...
```
Comment (borkaehw) [inline rules/scala.bzl]: Same concern here.

File: scripts/gen-deps.sh [hunk 3]
```diff
@@ -16,4 +16,4 @@ echo generating dependencies for tests workspace
 rm -fr tests/3rdparty
 external-tools/bazel-deps.sh generate -r "$(pwd)/tests" -s 3rdparty/maven.bzl -d dependencies.yaml
 
-./format.sh
+./scripts/format.sh
```
Comment (borkaehw) [inline scripts/gen-deps.sh]: I prefer to have separate PR for all these small fixes, same for gen docs. It's easier to track what has been changed later on.

File: rules/scala.bzl (hunk 1 above)
Comment (andyscott) [reply rules/scala.bzl]: The coverage phase is responsible for adding the jacoco instrumentation actions to the action graph.

File: src/main/scala/higherkindness/rules_scala/common/sbt-testing/CLassLoaders.scala [hunk 4]
```diff
@@ -15,6 +15,8 @@ object ClassLoaders {
   def sbtTestClassLoader(urls: Seq[URL]) = new URLClassLoader(urls.toArray, null) {
     private[this] val current = getClass.getClassLoader()
     override protected def findClass(className: String): Class[_] =
-      if (className.startsWith("sbt.testing.")) current.loadClass(className) else super.findClass(className)
+      if (className.startsWith("sbt.testing.")) current.loadClass(className)
+      else if (className.startsWith("org.jacoco.agent.rt.")) current.loadClass(className)
```
[the remark this answers is not in the document]
Comment (andyscott) [reply src/main/scala/higherkindness/rules_scala/common/sbt-testing/CLassLoaders.scala]: 👍

File: scripts/gen-deps.sh (hunk 3 above)
Comment (andyscott) [reply scripts/gen-deps.sh]: Done!

**Pull request merged**